### DISCUSSION TITLE
Commands/Operations Listing?

### DISCUSSION THREAD

**Discussion by DAWA48** [rusq/slackdump#191] [Q&A]
Is there a full list of supported commands available with or without descriptions? Perhaps I am not navigating the documentation well but I am learning a few other tips and tricks by looking though the comments and discussions. For example how does the "dump-from" function work?

Can you provide a synopsis on the different between the dump versus export functions as well please? What are the logic for using one versus another?

The overview posted on [medium.com](https://gilyazov.medium.com/downloading-your-private-slack-conversations-52e50428b3c2) mentions a low profile - but I imagine one can review API requests etc and see an uptick (or is it not a continuous pull)?. Can you provide more insight on that if you're willing.

**Comment by rusq** [comment]
Hey @DAWA48, thanks for highlighting the need for more doc.

You can get the list of available commands by running `./slackdump -h`, it provides the short description of each of the available flags.
You can also check the [Full CLI reference](https://github.com/rusq/slackdump/blob/master/doc/cli.rst)

For v2.2.x there are four major modes of operation:
- Dump — generate the API output dump as a JSON file
- List — list users and channels, can be used to save API output dump of users/channels.
- Export — generate (a compatible) Slack Export file
- Emoji — dump emojis for the workspace.

The main difference between Dump VS Export is that:
- **Dump** generates the files directly from API output (slight changes include: (a) threads are placed inside the messages that started them, (b) messages are sorted in ascending time order (VS descending, returned from the API)
- **Export** uses dump functions under the hood, but then applies transformations to the data to match the way that it is organised in the Slack Export file, namely (a) rearrange and save all messages by the date, (b) generate the index of Users, Channels, DMs and MPIMs in the root.  The main difference between the Slack Export generated by Slack, and Slack Export generated by Slackdump is that Slackdump downloads and saves the files locally, while Slack appends an export token suffix to each of the file URLs.  (Slackdump has an optional parameter that allows to append the token to the filenames too , see #105)

Depending on your need, you might choose one over another:
- Use Dump if:
  - you need to dump conversations for archival purposes (you can view the dump with [Slackdump2Html](https://github.com/kununu/slackdump2html))
  - if you want to generate the channel dump periodically, for example, if you're updating some database with the data from slack, and do it incrementally every day (that's where -dump-from and -dump-to would help)
  - etc
- Use Export if:
  - you need a compatibility with Slack Export format
  - need to create a snapshot of your slack workspace
  - need a full browsable archive of the workspace (and browse it with [SlackLogViewer](https://github.com/thayakawa-gh/SlackLogViewer) for example).

Regarding the "spikes" in API usage:  as far as I know the "analytics" that slack provides can't offer the detailed view on client token API request number, it provides just aggregated "messages sent/messages read", without the split by user,  but let me know if know otherwise.

**Comment by DAWA48** [comment]
Thanks for responding @rusq.

I guess it depends on the type of organisation. It looks like if it in an Enterprise grid there is an anomaly detection for "excessive downloads" https://api.slack.com/admins/audit-logs-anomaly which I presume can be kicked off by the Slackdump tool.

Otherwise from my understanding of what you are doing in the code, you are "scrapping" channels/threads and aggregating or recreating an export that somewhat mimics native functionality. Is that the correct understanding?

**Reply by rusq** [reply]
Thanks for the info on the anomaly detection, interesting!

Yes, that is correct.  It mimics the native archive format, with slight differences, described above.

**Reply by DAWA48** [reply]
I couldn't find much information on the parameters around the monitoring apis (what qualifies as excessive etc) but it's something to be aware of.

Since exports/dumps are just based on the view/read api results and formatted after the fact agreed it is likely low touch/risk as you've mentioned

**Comment by DAWA48** [comment]
@rusq can you point me to the files that have the api calls you are using please?

**Reply by rusq** [reply]
Sure,

these are the main ones:
- [conversations.history](https://github.com/rusq/slackdump/blob/master/messages.go#L114)
- [conversations.replies](https://github.com/rusq/slackdump/blob/master/thread.go#L117)
- [conversations.list](https://github.com/rusq/slackdump/blob/master/channels.go#L69)
- [users.list](https://github.com/rusq/slackdump/blob/master/users.go#L61)

This [interface](https://github.com/rusq/slackdump/blob/master/slackdump.go#L48-L55), has all the methods that slackdump uses, including some sundry ones.

What are you up to?
